oauthproxy.go

```diff
@@ -955,49 +956,72 @@ func (p *OAuthProxy) getAuthenticatedSession(rw http.ResponseWriter, req *http.R
 
 // authOnlyAuthorize handles special authorization logic that is only done
 // on the AuthOnly endpoint for use with Nginx subrequest architectures.
-//
-// TODO (@NickMeves): This method is a placeholder to be extended but currently
-// fails the linter. Remove the nolint when functionality expands.
-//
-//nolint:gosimple
 func authOnlyAuthorize(req *http.Request, s *sessionsapi.SessionState) bool {
-	// Allow secondary group restrictions based on the `allowed_groups`
-	// querystring parameter
-	if !checkAllowedGroups(req, s) {
-		return false
+	constraints := []func(*http.Request, *sessionsapi.SessionState) bool{
+		checkAllowedGroups,
+		checkAllowedEmailDomains,
+	}
+
+	for _, constraint := range constraints {
+		if !constraint(req, s) {
+			return false
+		}
 	}
 
 	return true
 }
 
-func checkAllowedGroups(req *http.Request, s *sessionsapi.SessionState) bool {
-	allowedGroups := extractAllowedGroups(req)
-	if len(allowedGroups) == 0 {
+func extractAllowedEntities(req *http.Request, field string) []string {
+	entities := []string{}
+
+	query := req.URL.Query()
+	for _, allowedEntities := range query[field] {
+		for _, entity := range strings.Split(allowedEntities, ",") {
+			if entity != "" {
+				entities = append(entities, entity)
+			}
+		}
+	}
+
+	return entities
+}
+
+// checkAllowedEmailDomains allow email domain restrictions based on the `allowed_email_domains`
+// querystring parameter
+func checkAllowedEmailDomains(req *http.Request, s *sessionsapi.SessionState) bool {
+	allowedEmailDomains := extractAllowedEntities(req, "allowed_email_domains")
+	if len(allowedEmailDomains) == 0 {
 		return true
 	}
 
-	for _, group := range s.Groups {
-		if _, ok := allowedGroups[group]; ok {
-			return true
-		}
+	emailSplited := strings.Split(s.Email, "@")
+	if len(emailSplited) != 2 {
+		return false
 	}
 
-	return false
+	endpoint, _ := url.Parse("")
+	endpoint.Host = emailSplited[1]
+
+	return util.IsEndpointAllowed(endpoint, allowedEmailDomains)
 }
 
-func extractAllowedGroups(req *http.Request) map[string]struct{} {
-	groups := map[string]struct{}{}
+// checkAllowedGroups allow secondary group restrictions based on the `allowed_groups`
+// querystring parameter
+func checkAllowedGroups(req *http.Request, s *sessionsapi.SessionState) bool {
+	allowedGroups := extractAllowedEntities(req, "allowed_groups")
+	if len(allowedGroups) == 0 {
+		return true
+	}
 
-	query := req.URL.Query()
-	for _, allowedGroups := range query["allowed_groups"] {
-		for _, group := range strings.Split(allowedGroups, ",") {
-			if group != "" {
-				groups[group] = struct{}{}
+	for _, group := range s.Groups {
```

Comment:
What's the reasoning for moving away from Loop + Set (with `O(NlogN)` time complexity) to double looping to find matches (`O(N^2)` time complexity?

I know in small cases (< 10), double looping is fine and potentially faster. But since we don't know how large our worst case users will be using this feature -- we errored toward `O(NlogN)` Loop + Set in the initial implementation.

Reply:
I recommend switching the `extractAllowedEntities` implementation to return a `map[string]struct{}` instead of `[]string` and doing a set membership check in your checker methods.

Reply:
Yes you right, I fixed it with the recommended implementation.